lib/cea/cea608_memory.js

```diff
@@ -0,0 +1,513 @@
+   * @param {!string} bgColor
+   * @return {!shaka.text.Cue}
+   * @private
+   */
+  createStyledCue_(startTime, endTime, text, underline, italics,
+      txtColor, bgColor) {
+    const cue = new shaka.text.Cue(startTime, endTime, text);
+    if (underline) {
+      cue.textDecoration.push(shaka.text.Cue.textDecoration.UNDERLINE);
+    }
+    if (italics) {
+      cue.fontStyle = shaka.text.Cue.fontStyle.ITALIC;
+    }
+    cue.color = txtColor;
+    cue.backgroundColor = bgColor;
+    return cue;
+  }
+
+  /**
+   * Resets the memory buffer.
+   */
+  reset() {
+    this.resetAllRows();
+    this.row_ = 1;
+  }
+
+  /**
+   * @return {!number}
+   */
+  getRow() {
+    return this.row_;
+  }
+
+  /**
+   * @param {!number} row
+   */
+  setRow(row) {
+    this.row_ = row;
+  }
+
+  /**
+   * @return {!number}
+   */
+  getScrollSize() {
+    return this.scrollRows_;
+  }
+
+  /**
+   * @param {!number} scrollRows
+   */
+  setScrollSize(scrollRows) {
+    this.scrollRows_ = scrollRows;
+  }
+
+  /**
+   * Adds a character to the buffer.
+   * @param {!shaka.cea.Cea608Memory.CharSet} set Character set.
+   * @param {!number} b CC byte to add.
+   */
+  addChar(set, b) {
+    // Valid chars are in the range [0x20, 0x7f]
+    if (b < 0x20 || b > 0x7f) {
+      return;
+    }
+
+    let char = '';
+    switch (set) {
+      case shaka.cea.Cea608Memory.CharSet.BASIC_NORTH_AMERICAN:
+        if (shaka.cea.Cea608Memory.CharSet.BasicNorthAmericanChars.has(b)) {
+          char =
+                shaka.cea.Cea608Memory.CharSet.BasicNorthAmericanChars.get(b);
+        } else {
+          // Regular ASCII
+          char = String.fromCharCode(b);
+        }
+        break;
+      case shaka.cea.Cea608Memory.CharSet.SPECIAL_NORTH_AMERICAN:
+        char =
+              shaka.cea.Cea608Memory.CharSet.SpecialNorthAmericanChars.get(b);
+        break;
+      case shaka.cea.Cea608Memory.CharSet.SPANISH_FRENCH:
+        // Extended charset does a BS over preceding char, 6.4.2 EIA-608-B.
+        this.eraseChar();
+        char =
+              shaka.cea.Cea608Memory.CharSet.ExtendedSpanishFrench.get(b);
+        break;
+      case shaka.cea.Cea608Memory.CharSet.PORTUGUESE_GERMAN:
+        this.eraseChar();
+        char =
+              shaka.cea.Cea608Memory.CharSet.ExtendedPortugueseGerman.get(b);
+        break;
+    }
+
+    if (char) {
+      const styledChar = new shaka.cea.Cea608Char(char, this.underline_,
+          this.italics_, this.backgroundColor_, this.textColor_);
+      this.rows_[this.row_].push(styledChar);
+    }
+  }
+
+  /**
+   * Erases a character from the buffer.
+   */
+  eraseChar() {
+    this.rows_[this.row_].pop();
+  }
+
+  /**
+   * Moves rows of characters.
+   * @param {!number} dst Destination row index.
+   * @param {!number} src Source row index.
+   * @param {!number} count Count of rows to move.
+   */
+  moveRows(dst, src, count) {
+    for (let i = 0; i < count; i++) {
+      this.rows_[dst + i] = this.rows_[src + i];
```

Comment:
Arrays in JavaScript are references, so this will result in two rows in the rows_ array being the same object.  For example, it you call moveRows(1, 0, 1), now rows_[0].pop() affects both rows_[0] and rows_[1] because they are literally the same object.

I see in Cea608DataChannel that the caller tends to call resetRows() right afterward to reset the rows that got moved.  Is that a requirement?  If so, it should be documented in the description of moveRows().

Also, if (dst - src) < count, it looks like you end up destroying data.  For example:

```js
dst = 1;
src = 0;
count = 2;
rows[1] = rows[0];
rows[2] = rows[1];
```

Now the original data in rows[1] was lost and both rows got a copy of the original rows[0].

Must dst always be larger than src?  If so, you could iterate backward to solve that issue.  If not, the iteration direction depends on the sign of (dst-src).

Reply:
This is a really good catch. As for the first issue of references, I've used a mapper to clone the array.
As for the second issue, I've done what you said and iterate backwards on dst>=src, and forwards otherwise.